node/libs/storage/src/block_store/mod.rs

```diff
@@ -0,0 +1,246 @@
+        static COLLECTOR: vise::Collector<Option<metrics::BlockStore>> = vise::Collector::new();
+        let store_ref = Arc::downgrade(&self.0);
+        let _ = COLLECTOR.before_scrape(move || Some(store_ref.upgrade()?.scrape_metrics()));
+
+        let res = async {
+            let inner = &mut self.0.inner.subscribe();
+            loop {
+                let block = sync::wait_for(ctx, inner, |inner| !inner.queue.is_empty())
+                    .await?
+                    .queue[0]
+                    .clone();
+
+                // TODO: monitor errors as well.
+                let t = metrics::PERSISTENT_BLOCK_STORE
+                    .store_next_block_latency
+                    .start();
+                self.0.persistent.store_next_block(ctx, &block).await?;
+                t.observe();
+
+                self.0.inner.send_modify(|inner| {
+                    debug_assert_eq!(inner.persisted_state.next(), block.header().number);
+                    inner.persisted_state.last = block.justification.clone();
+                    inner.queue.pop_front();
+                });
+            }
+        }
+        .await;
+        match res {
+            Ok(()) | Err(ctx::Error::Canceled(_)) => Ok(()),
+            Err(ctx::Error::Internal(err)) => Err(err),
+        }
+    }
+}
+
+impl BlockStore {
+    /// Constructs a BlockStore.
+    /// BlockStore takes ownership of the passed PersistentBlockStore,
+    /// i.e. caller should modify the underlying persistent storage
+    /// ONLY through the constructed BlockStore.
+    pub async fn new(
+        ctx: &ctx::Ctx,
+        persistent: Box<dyn PersistentBlockStore>,
+    ) -> ctx::Result<(Arc<Self>, BlockStoreRunner)> {
+        let t = metrics::PERSISTENT_BLOCK_STORE.state_latency.start();
+        let state = persistent.state(ctx).await.wrap("persistent.state()")?;
+        t.observe();
+        if state.first.header().number > state.last.header().number {
+            return Err(anyhow::anyhow!("invalid state").into());
+        }
+        let this = Arc::new(Self {
+            persistent,
+            inner: sync::watch::channel(Inner {
+                queued_state: sync::watch::channel(state.clone()).0,
+                persisted_state: state,
+                queue: VecDeque::new(),
+            })
+            .0,
+        });
+        Ok((this.clone(), BlockStoreRunner(this)))
+    }
+
+    /// Fetches a block (from queue or persistent storage).
+    pub async fn block(
+        &self,
+        ctx: &ctx::Ctx,
+        number: validator::BlockNumber,
+    ) -> ctx::Result<Option<validator::FinalBlock>> {
+        {
+            let inner = self.inner.borrow();
+            if !inner.queued_state.borrow().contains(number) {
+                return Ok(None);
+            }
+            if !inner.persisted_state.contains(number) {
+                let idx = number.0 - inner.persisted_state.next().0;
+                return Ok(Some(inner.queue[idx as usize].clone()));
```

Comment:
Nit: I'd expect checked subtraction and `get(_)` instead of indexing `queue`; or a comment why these ops are safe (because `inner.queued_state.borrow().contains(number)` was checked earlier, right?).

Reply:
done